### PR TITLE
RUMM-2938 Create DatadogInternal Module

### PR DESCRIPTION
**Pull request by maxep**
### What and why?

Create the `DatadogInternal` module that will be shared will all Datadog modules. It includes Interface definitions and utilities to build and Core and Features.

> Sorry for the large PR, Xcode targets contains most of the added lines.

### How?

The `DatadogInternal` module is created as both a SPM target and as a Xcode target while doing the migration. Access levels have been updated accordingly 

### Review checklist
- [x] Feature or bugfix MUST have appropriate tests (unit, integration)
- [x] Make sure each commit and the PR mention the Issue number or JIRA reference
- [ ] Add CHANGELOG entry for user facing changes

### Custom CI job configuration (optional)
- [x] Run unit tests
- [x] Run integration tests
- [x] Run smoke tests

### CODE REVIEW

File: DatadogInternal/Sources/Context/DatadogSite.swift
Comment (maxep) [inline DatadogInternal/Sources/Context/DatadogSite.swift:9]: Needs to be updated with #1157

File: DatadogInternal/Sources/DD.swift
Comment (maxep) [inline DatadogInternal/Sources/DD.swift:24]: It is now set to `debug` by default, it is then changed after SDK init.

File: DatadogInternal/Sources/Telemetry/InternalLogger.swift
Comment (maxep) [inline DatadogInternal/Sources/Telemetry/InternalLogger.swift:21]: It no longer use the Logs Feature levels.

File: DatadogInternal/Sources/Extensions/DatadogExtended.swift
Comment (maxep) [inline DatadogInternal/Sources/Extensions/DatadogExtended.swift:51]: This is renamed to no conflict with the package name.

File: Sources/Datadog/DatadogConfiguration.swift
Comment (maxep) [inline Sources/Datadog/DatadogConfiguration.swift:51]: Moved to `DatadogInternal` as `DatadogSite`.

File: Tests/DatadogTests/Datadog/DatadogTests.swift
Comment (maxep) [inline Tests/DatadogTests/Datadog/DatadogTests.swift:109]: I haven't found a replacement yet :/

File: tools/rum-models-generator/Sources/rum-models-generator/GenerateRUMModels.swift
Comment (maxep) [inline tools/rum-models-generator/Sources/rum-models-generator/GenerateRUMModels.swift:22]: The model needs the `DynamicCodingKey` from the internal module.

File: DatadogInternal/Sources/Context/DatadogSite.swift
Comment (maxep) [reply DatadogInternal/Sources/Context/DatadogSite.swift:9]: Done.

File: Datadog/Example/AppConfiguration.swift
Comment (ncreated) [inline Datadog/Example/AppConfiguration.swift:9]: Why do we need to import `DatadogInternal` in application code? In theory, `DatadogInternal` shouldn't be used by end-users. Is this required by migration some step or Example project specifics?

File: DatadogInternal/Sources/Upload/DataCompression.swift
Comment (ncreated) [inline DatadogInternal/Sources/Upload/DataCompression.swift:19]: question/ Do we need compression things in public interface? The `URLRequestBuilder` handles compression:
```swift
// URLRequestBuilder.swift

func uploadRequest(with body: Data, compress: Bool = true)
```
so details of the actual `encode()`, `compress()` and `addler32()` could be internal, no?

File: Sources/Datadog/DatadogCore/DatadogCore.swift
Comment (ncreated) [inline Sources/Datadog/DatadogCore/DatadogCore.swift:508]: suggestion/ Defaults resolution is error-prone and dangerous if done in different places. For that reason, in V1 convention was to resolve defaults early in configuration declaration - so we don't need to struggle with optionals in downstream logic.

I think the root cause of optional here is:
```swift
/// [Datadog Site](https://docs.datadoghq.com/getting_started/site/) for data uploads. It can be `nil` in V1
/// if the SDK is configured using deprecated APIs: `set(logsEndpoint:)`, `set(tracesEndpoint:)` and `set(rumEndpoint:)`.
let site: DatadogSite?
```
I'd recommend tackling this legacy early and individually (isolated PR would be best). If we remove these deprecated APIs on `v2` branch, it will streamline further changes around `site` logic.

File: DatadogInternal/Sources/Telemetry/Telemetry.swift
Comment (ncreated) [inline DatadogInternal/Sources/Telemetry/Telemetry.swift:32]: question/ How does configuration telemetry work now? I struggled with finding corresponding changes, only noticed disabled test 🤔💭.

**Review by ncreated — COMMENTED** [review]
Extreme efforts, thanks @maxep 🏋️‍♂️🏅.

A bit hard to navigate in this PR. I can see diffs from few other PRs as well (SR bitmaps, `ap1()` DC, etc.) - I assume it comes from `→ develop` update. If possible, let's try smaller changes in next migration work 🙏 as it visibly impacts the quality of CR I can give here.

I left few questions to clarify some elements before proceeding with merge. But it looks great in overall 🚀!

**Comment by maxep** [comment]
Thanks for the review @ncreated 🙏 

Yes, `feature/v2` needs an update with `develop`, I'm currently rebasing it so this PR will get a bit smaller.

File: Datadog/Example/AppConfiguration.swift
Comment (maxep) [reply Datadog/Example/AppConfiguration.swift:9]: Ah good catch! we need to expose some types from `DatadogInternal`, but this should be transparent for the user.
We can either use wrapper types or use the `@_exported import` for some types. That second option is easier!

File: Sources/Datadog/DatadogCore/DatadogCore.swift
Comment (maxep) [reply Sources/Datadog/DatadogCore/DatadogCore.swift:508]: Good point, I agree 👍

File: Datadog/Example/AppConfiguration.swift
Comment (ncreated) [reply Datadog/Example/AppConfiguration.swift:9]: > we need to expose some types from DatadogInternal, but this should be transparent for the user.

Which ones for example? Yes, I think `@_exported import` in our `Example` project should do the job 👍.

File: DatadogInternal/Sources/Telemetry/Telemetry.swift
Comment (maxep) [reply DatadogInternal/Sources/Telemetry/Telemetry.swift:32]: The `Telemetry` protocol does not require this method, but the `RUMTelemetry` does expose it. The reason being that the `FeaturesConfiguration` type is not part of the `DatadogInternal`, it's define in the `Datadog` module as well as `RUMTelemetry`. [The configuration is still reported at initialisation](https://github.com/DataDog/dd-sdk-ios/pull/1154/files#diff-7fae8d57fd7c8b0dc28f762018d1019d1f75530a064e083e7828d309e182be3dR299).

It's still unclear how we will tackle reporting configuration telemetry, but it will be tackled separately.

File: Datadog/Example/AppConfiguration.swift
Comment (maxep) [reply Datadog/Example/AppConfiguration.swift:9]: Currently, I see the following (it needs to be defined in `Datadog` module)

```swift
@_exported import enum DatadogInternal.DatadogSite
@_exported import struct DatadogInternal.AttributeKey
@_exported import protocol DatadogInternal.AttributeValue
@_exported import enum DatadogInternal.TrackingConsent
```

**Note:** I'm thinking to remove `AttributeKey` and `AttributeValue` aliases at some point

Comment (ncreated) [reply Datadog/Example/AppConfiguration.swift]: Ok, now it makes sense 👌 - we need these symbols in `Example` app because we add custom logics around it (e.g. use it in variables declaration). Such use case won't happen in customer app, because the symbol will be visible through `Datadog` or other Feature modules.

> Note: I'm thinking to remove AttributeKey and AttributeValue aliases at some point

Their role today is to [describe with code comments](https://github.com/DataDog/dd-sdk-ios/blob/5491d29171e08c939ee09cc0d6e51ad66309daa6/Sources/Datadog/Core/Attributes/Attributes.swift#L9-L63) how keys and values work across all products. If we remove these aliases, where would we describe it? The only place would be next to each method that depends on attributes, which would be tedious. Not sure why we'd want to remove them?

File: Sources/Datadog/DatadogConfiguration.swift
Comment (ncreated) [inline Sources/Datadog/DatadogConfiguration.swift:12]: I was thinking we will do it in `Example` app project - do it once (in `AppDelegate`?) and have these symbols re-exported to the entire app module 🤔💭. This could be minimal and isolated solution to our inconvenience.

Unlike what we seek for in `Example` app, our users don't need to have these types available for variables declaration, e.g.:
```swift
let trackingConent: TrackingConsent
```
They will only use it for variables assignment:
```swift
.set(trackingConsent:)
```
so we don't need to re-export their visibility.

Also, `@_exported` is discouraged by Swift maintainers, same as other [underscored attributes](https://github.com/apple/swift/blob/main/docs/ReferenceGuides/UnderscoredAttributes.md). Better to avoid it if possible 😉. Here, it's perfectly possible as the problem is only in our `Example` project. WDYT @maxep ?

File: Datadog/Example/AppConfiguration.swift
Comment (maxep) [reply Datadog/Example/AppConfiguration.swift]: Yes, the doc well explains it, but it is necessary because `[AttributeKey: AttributeValue]` is kindof confusing. While `[String: Encodable]` is clearer IMO. We could use a typealias like 
```swift
typealias DatadogAdditionalAttributes = [String: Encodable]
```

**Review by ncreated — APPROVED** [review]
Looks great! I only left one [feedback](https://github.com/DataDog/dd-sdk-ios/pull/1154/commits/b8b78e2c6ccb2ecafbc62b5bce0fa749492ff5e5#r1105517772) on `@_exported import` (please have a look). Other than this, good to go, excellent 🚀!

File: Datadog/Example/AppConfiguration.swift
Comment (ncreated) [reply Datadog/Example/AppConfiguration.swift]: `DatadogAdditionalAttributes = [String: Encodable]` could work as well 👍. I like type aliases because they have some benefits of extra types without requiring us to introduce new `structs` in fact. Let's keep this for later, agree 👍.

File: Sources/Datadog/DatadogConfiguration.swift
Comment (maxep) [reply Sources/Datadog/DatadogConfiguration.swift:12]: I think we only need the `TrackingConsent` in the Example, I will double check.

But it's very likely that users will need it as well, to implement any consent logic, e.g. storing in UserDefault or map the type with a Consent Provider. So we would probably want to expose it somehow, no?

I saw that the underscored attributes are discouraged but I don't see the risk with this one, other than semantic change. But to avoid any, the other option is having a wrapper exposed by the `Datadog` module:
```swift
public enum DatadogTrackingConsent {
    ...
    internal var _internal: DatatogInternal.TrackingConsent { get }
}
```
I will go with that option, if that' ok with you?

Comment (maxep) [reply Sources/Datadog/DatadogConfiguration.swift:12]: Well, the wrapper has a larger impact than I thought :/ I will go with your suggestion to move forward 👍

Comment (ncreated) [reply Sources/Datadog/DatadogConfiguration.swift]: > But it's very likely that users will need it as well, to implement any consent logic, e.g. storing in UserDefault or map the type with a Consent Provider. So we would probably want to expose it somehow, no?

Could be 👌, but this is quite an assumption that we have no proof for. I'd say let's keep it simple for now - if one asks for it, we will think of solution. We could always add thin wrapper just for SDK configuration part (in `Datadog`) that maps into `DatadogInternal.TrackingConsent` right after configuration (with no other use than configuration it will have minimal to no impact).

**Pull request merged**
